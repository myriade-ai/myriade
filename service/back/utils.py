from typing import Any, Dict, List, Optional
from uuid import UUID

from flask import g

from models import Database
from models.catalog import Asset, ColumnFacet, TableFacet


def create_database(
    name: str,
    description: str,
    engine: str,
    details: dict,
    public: bool,
    write_mode: str = "confirmation",
    owner_id: str | None = None,
    organisation_id: str | None = None,
    dbt_catalog: dict | None = None,
    dbt_manifest: dict | None = None,
    dbt_repo_path: str | None = None,
):
    # Create a new database
    database = Database(
        name=name,
        description=description,
        engine=engine,
        details=details,
        organisationId=organisation_id,
        ownerId=owner_id,
        public=public,
        write_mode=write_mode,
        dbt_catalog=dbt_catalog,
        dbt_manifest=dbt_manifest,
        dbt_repo_path=dbt_repo_path,
    )
    # After database creation, we'll sync initial metadata to catalog
    # This happens in the API layer after the database is persisted
    return database


def merge_tables_metadata(
    existing: Optional[List[Dict[str, Any]]], new: List[Dict[str, Any]]
) -> List[Dict[str, Any]]:
    """Return `new` tables list enriched with privacy maps from `existing`.

    If a table/column already has a privacy map in `existing`, keep it.
    Tables not present in the data_warehouse anymore are discarded.
    """

    existing_lookup = {}
    if existing:
        for t in existing:
            existing_lookup[(t.get("schema"), t["name"])] = t

    merged: List[Dict[str, Any]] = []
    for t in new:
        key = (t.get("schema"), t["name"])
        previous = existing_lookup.get(key)
        # Build columns preserving privacy if available
        merged_columns: List[Dict[str, Any]] = []
        for col in t["columns"]:
            # find same column in previous
            prev_col_priv = {}
            if previous:
                for pc in previous.get("columns", []):
                    if pc["name"] == col["name"]:
                        prev_col_priv = pc.get("privacy", {})
                        break
            merged_columns.append(
                {
                    **col,
                    "privacy": prev_col_priv.copy(),
                }
            )
        merged.append({**t, "columns": merged_columns})

    return merged


def get_tables_metadata_from_catalog(
    database_id: UUID, session=None
) -> List[Dict[str, Any]]:
    """Get tables metadata from catalog in the same format as the old
    tables_metadata field

    Args:
        database_id: UUID of the database
        session: Optional SQLAlchemy session. If not provided, uses g.session
    """
    # Use provided session or fall back to g.session
    db_session = session if session is not None else g.session

    try:
        # Get all table and column assets for this database
        tables_data = (
            db_session.query(Asset, TableFacet)
            .join(TableFacet, Asset.id == TableFacet.asset_id)
            .filter(Asset.database_id == database_id, Asset.type == "TABLE")
            .all()
        )

        columns_data = (
            db_session.query(Asset, ColumnFacet, TableFacet)
            .join(ColumnFacet, Asset.id == ColumnFacet.asset_id)
            .join(TableFacet, ColumnFacet.parent_table_asset_id == TableFacet.asset_id)
            .filter(Asset.database_id == database_id, Asset.type == "COLUMN")
            .order_by(ColumnFacet.ordinal)
            .all()
        )

        # Group columns by table
        tables_map = {}
        for asset, table_facet in tables_data:
            table_key = (table_facet.schema, table_facet.table_name)
            tables_map[table_key] = {
                "name": table_facet.table_name,
                "schema": table_facet.schema,
                "description": asset.description,
                "table_type": table_facet.table_type,
                "columns": [],
            }

        # Add columns to their tables
        for asset, column_facet, table_facet in columns_data:
            table_key = (table_facet.schema, table_facet.table_name)
            if table_key in tables_map:
                column_info = {
                    "name": column_facet.column_name,
                    "type": column_facet.data_type,
                    "description": asset.description,
                    "privacy": column_facet.privacy or {},
                }
                tables_map[table_key]["columns"].append(column_info)

        return list(tables_map.values())

    except Exception:
        # If catalog is not available or fails, rollback and return empty list
        db_session.rollback()
        return []


def sync_database_metadata_to_assets(
    database_id: UUID,
    tables_metadata: List[Dict[str, Any]],
    session=None,
):
    """
    Sync catalog with database metadata (tables/columns)
    This method creates catalog assets from the database's tables_metadata
    using facet-based models. It also removes assets that no longer exist
    in the database.

    Args:
        database_id: UUID of the database
        tables_metadata: List of table metadata dictionaries
        session: Optional SQLAlchemy session. If not provided, uses g.session
    """
    if not tables_metadata:
        raise ValueError("No database metadata available to sync")

    # Use provided session or fall back to g.session
    db_session = session if session is not None else g.session

    synced_count = 0

    # Track all valid URNs from current metadata (both tables and columns)
    valid_urns = set()

    for table_meta in tables_metadata:
        schema_name = table_meta.get("schema")
        table_name = table_meta.get("name")

        # Generate URN for table
        table_urn = f"urn:table:{database_id}:{schema_name}:{table_name}"
        valid_urns.add(table_urn)

        # Check if table asset already exists
        table_asset = (
            db_session.query(Asset)
            .filter(
                Asset.database_id == database_id,
                Asset.urn == table_urn,
            )
            .first()
        )

        if not table_asset:
            # Create new table asset
            table_asset = Asset(
                name=table_name,
                urn=table_urn,
                type="TABLE",
                description=table_meta.get("description"),
                database_id=database_id,
            )
            db_session.add(table_asset)
            db_session.flush()  # Get the ID

            # Create table facet
            table_facet = TableFacet(
                asset_id=table_asset.id,
                database_id=database_id,
                schema=schema_name,
                table_name=table_name,
                table_type=table_meta.get("table_type"),
            )
            db_session.add(table_facet)
            synced_count += 1
        else:
            # Update existing table facet's table_type field if it changed
            table_facet = (
                db_session.query(TableFacet)
                .filter(TableFacet.asset_id == table_asset.id)
                .first()
            )
            if table_facet:
                table_facet.table_type = table_meta.get("table_type").upper()

        # Create/update column assets
        for i, column in enumerate(table_meta.get("columns", [])):
            column_name = column.get("name")

            # Generate URN for column
            column_urn = (
                f"urn:column:{database_id}:{schema_name}:{table_name}:{column_name}"
            )
            valid_urns.add(column_urn)

            column_asset = (
                db_session.query(Asset)
                .filter(
                    Asset.database_id == database_id,
                    Asset.urn == column_urn,
                )
                .first()
            )

            if not column_asset:
                # Create new column asset
                column_asset = Asset(
                    name=column_name,
                    urn=column_urn,
                    type="COLUMN",
                    description=column.get("description"),
                    database_id=database_id,
                )
                db_session.add(column_asset)
                db_session.flush()  # Get the ID

                # Create column facet
                column_facet = ColumnFacet(
                    asset_id=column_asset.id,
                    parent_table_asset_id=table_asset.id,
                    column_name=column_name,
                    ordinal=i,
                    data_type=column.get("type"),
                    privacy=column.get("privacy"),
                )
                db_session.add(column_facet)
                synced_count += 1

    # Delete assets that no longer exist in the database
    orphaned_assets = (
        db_session.query(Asset)
        .filter(
            Asset.database_id == database_id,
            ~Asset.urn.in_(valid_urns) if valid_urns else True,
        )
        .all()
    )

    for asset in orphaned_assets:
        if asset.type == "COLUMN":
            db_session.query(ColumnFacet).filter(
                ColumnFacet.asset_id == asset.id
            ).delete()
        elif asset.type == "TABLE":
            db_session.query(TableFacet).filter(
                TableFacet.asset_id == asset.id
            ).delete()
        db_session.delete(asset)

    return {
        "synced_count": synced_count,
    }


def update_catalog_privacy(
    database_id: UUID, tables_metadata: List[Dict[str, Any]], session=None
):
    """
    Update privacy settings in catalog for existing assets
    This method updates column facets with new privacy information

    Args:
        database_id: UUID of the database
        tables_metadata: List of table metadata dictionaries
        session: Optional SQLAlchemy session. If not provided, uses g.session
    """
    if not tables_metadata:
        return

    db_session = session if session is not None else g.session

    for table_meta in tables_metadata:
        schema_name = table_meta.get("schema")
        table_name = table_meta.get("name")

        # Update column privacy settings
        for column in table_meta.get("columns", []):
            column_name = column.get("name")
            privacy_settings = column.get("privacy", {})

            # Generate URN for column
            column_urn = (
                f"urn:column:{database_id}:{schema_name}:{table_name}:{column_name}"
            )

            # Find existing column asset
            column_asset = (
                db_session.query(Asset)
                .filter(
                    Asset.database_id == database_id,
                    Asset.urn == column_urn,
                    Asset.type == "COLUMN",
                )
                .first()
            )

            if column_asset:
                # Update column facet privacy settings
                column_facet = (
                    db_session.query(ColumnFacet)
                    .filter(ColumnFacet.asset_id == column_asset.id)
                    .first()
                )

                if column_facet:
                    column_facet.privacy = privacy_settings
