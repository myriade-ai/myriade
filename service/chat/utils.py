import re


def parse_answer_text(text: str):
    """Parse the text in a list of chunks.
    > parse_answer_text("Hello <QUERY:1> and <CHART:2>")
    [{type: "text", content: "Hello"},
     {type: "query", query_id: 1},
     {type: "text", content: " and "},
     {type: "chart", chart_id: 2}]
    """
    if not text:
        return []

    chunks = []
    # Regex to find <QUERY:id> or <CHART:id> tags and capture the id
    tag_regex = r"(<QUERY:([^>]+)>)|(<CHART:([^>]+)>)"

    last_end = 0
    for match in re.finditer(tag_regex, text):
        start, end = match.span()

        # Add preceding text segment if it exists
        if start > last_end:
            chunks.append({"type": "text", "content": text[last_end:start]})

        # Check which group matched to determine type and get the ID
        if match.group(1):  # Matched <QUERY:id>
            query_id = match.group(2).strip()
            chunks.append({"type": "query", "query_id": query_id})
        elif match.group(3):  # Matched <CHART:id>
            chart_id = match.group(4).strip()
            chunks.append({"type": "chart", "chart_id": chart_id})

        last_end = end

    # Add any remaining text segment after the last tag
    if last_end < len(text):
        chunks.append({"type": "text", "content": text[last_end:]})

    return chunks
