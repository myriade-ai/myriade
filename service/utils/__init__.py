"""Utility modules for Myriade service"""
