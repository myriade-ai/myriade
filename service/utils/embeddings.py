"""
Lightweight embedding utility for vector search.

Uses sentence-transformers with a small model (all-MiniLM-L6-v2) for generating
embeddings. The model is ~80MB and provides a good balance between size and quality.

This module handles lazy loading of the model and provides a simple interface for
generating embeddings from text.
"""

import logging
from typing import List, Optional

logger = logging.getLogger(__name__)

# Global model instance (lazy-loaded)
_embedding_model = None


def get_embedding_model():
    """
    Get or initialize the sentence-transformers model.
    
    Uses lazy loading to avoid loading the model until it's actually needed.
    This is important for:
    - Faster application startup
    - Avoiding model loading in SQLite environments where vector search isn't available
    """
    global _embedding_model
    
    if _embedding_model is None:
        try:
            from sentence_transformers import SentenceTransformer
            
            logger.info("Loading embedding model: all-MiniLM-L6-v2")
            _embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
            logger.info("Embedding model loaded successfully")
        except ImportError:
            logger.warning(
                "sentence-transformers not installed. Vector search will not be available. "
                "Install with: pip install sentence-transformers"
            )
            _embedding_model = None
        except Exception as e:
            logger.error(f"Failed to load embedding model: {e}")
            _embedding_model = None
    
    return _embedding_model


def generate_embedding(text: str) -> Optional[List[float]]:
    """
    Generate an embedding vector for the given text.
    
    Args:
        text: Input text to embed
        
    Returns:
        List of floats representing the embedding vector (384 dimensions),
        or None if embedding generation fails or model is not available.
    """
    if not text or not text.strip():
        return None
    
    model = get_embedding_model()
    if model is None:
        return None
    
    try:
        # Generate embedding
        embedding = model.encode(text, convert_to_numpy=True)
        # Convert to list for JSON serialization
        return embedding.tolist()
    except Exception as e:
        logger.error(f"Failed to generate embedding: {e}")
        return None


def generate_asset_embedding(asset_name: str, asset_description: Optional[str] = None) -> Optional[List[float]]:
    """
    Generate an embedding for a catalog asset.
    
    Combines the asset name and description (if available) to create a more
    comprehensive embedding that captures both the asset identifier and its meaning.
    
    Args:
        asset_name: Name/URN of the asset
        asset_description: Optional description of the asset
        
    Returns:
        Embedding vector or None if generation fails
    """
    # Combine name and description for better semantic representation
    text_parts = [asset_name]
    
    if asset_description and asset_description.strip():
        text_parts.append(asset_description)
    
    combined_text = " ".join(text_parts)
    return generate_embedding(combined_text)
