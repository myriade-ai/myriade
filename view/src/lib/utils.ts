import type { ClassValue } from 'clsx'
import { clsx } from 'clsx'
import { format } from 'sql-formatter'
import { twMerge } from 'tailwind-merge'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export const formatSQL = (sql: string) => {
  return format(sql, { language: 'postgresql' })
}
