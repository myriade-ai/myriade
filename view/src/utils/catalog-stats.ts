import type { CatalogAsset } from '@/stores/catalog'

interface CatalogStats {
  total_assets: number
  completion_score: number
  assets_to_review: number
  assets_validated: number
  assets_with_ai_suggestions: number
  assets_with_description: number
}

/**
 * Compute catalog statistics from an array of assets
 * This allows calculating stats at any level (database, schema, table)
 */
export function computeCatalogStats(assets: CatalogAsset[]): CatalogStats {
  const totalAssets = assets.length

  if (totalAssets === 0) {
    return {
      total_assets: 0,
      completion_score: 0,
      assets_to_review: 0,
      assets_validated: 0,
      assets_with_ai_suggestions: 0,
      assets_with_description: 0
    }
  }

  const assetsWithDescription = assets.filter(
    (asset) => asset.description && asset.description.trim()
  ).length

  const assetsToReview = assets.filter(
    (asset) => asset.status === 'needs_review' || asset.status === 'requires_validation'
  ).length

  const assetsValidated = assets.filter((asset) => asset.status === 'validated').length

  const assetsWithAiSuggestions = assets.filter((asset) => asset.ai_suggestion).length

  const completionScore = Math.round((assetsWithDescription / totalAssets) * 100 * 10) / 10

  return {
    total_assets: totalAssets,
    completion_score: completionScore,
    assets_to_review: assetsToReview,
    assets_validated: assetsValidated,
    assets_with_ai_suggestions: assetsWithAiSuggestions,
    assets_with_description: assetsWithDescription
  }
}
