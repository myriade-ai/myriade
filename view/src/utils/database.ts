export interface Database {
  id: string
  createdAt: Date
  name: string
  engine: string
  details: any
  safe_mode: boolean
  dbt_catalog: any
  dbt_manifest: any
}

export const makeEmptyDatabase = () => ({
  id: '',
  createdAt: new Date(),
  name: '',
  engine: '',
  details: null,
  safe_mode: true,
  dbt_catalog: null,
  dbt_manifest: null
})

export const getDefaultDetailsForEngine = (engine: string) => {
  switch (engine) {
    case 'postgres':
      return { host: '', port: 5432, user: '', password: '', database: '' }
    case 'mysql':
      return { host: '', port: 3306, user: '', password: '', database: '' }
    case 'snowflake':
      return {
        account: '',
        user: '',
        password: '',
        database: '',
        schema: '',
        role: '',
        warehouse: ''
      }
    case 'sqlite':
      return { filename: '' }
    case 'bigquery':
      return { project_id: '', service_account_json: null }
    default:
      throw new Error('Unsupported engine')
  }
}

export const getDatabaseTypeName = (engine: string) => {
  const engineNames: Record<string, string> = {
    postgres: 'PostgreSQL',
    mysql: 'MySQL',
    snowflake: 'Snowflake',
    sqlite: 'SQLite',
    bigquery: 'BigQuery'
  }
  return engineNames[engine] || engine
}
