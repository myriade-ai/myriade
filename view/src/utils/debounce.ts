/**
 * Creates a debounced function that delays execution until after the specified wait time has elapsed
 * @param func - The function to debounce
 * @param wait - The delay in milliseconds
 * @returns A debounced function
 */
export function debounce<T extends (...args: any[]) => any>(
  func: T,
  wait: number
): (...args: Parameters<T>) => void {
  let timeoutId: ReturnType<typeof setTimeout> | null = null

  return function (...args: Parameters<T>) {
    if (timeoutId !== null) {
      clearTimeout(timeoutId)
    }
    timeoutId = setTimeout(() => {
      func(...args)
    }, wait)
  }
}
