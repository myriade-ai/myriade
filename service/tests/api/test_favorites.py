import requests

from tests.utils import normalise_json

created_query_id_for_favorite_test = None


def test_create_query_for_favorite_testing(app_server, test_db_id, snapshot):
    """Create a query to be used in favorite tests."""
    global created_query_id_for_favorite_test
    query_payload = {
        "title": "Test Query for Favorites",
        "sql": "SELECT 100 AS fav_col;",
        "databaseId": test_db_id,
    }
    response = requests.post(
        f"{app_server}/query",  # Endpoint to create a persistent query object
        json=query_payload,
        cookies={"wos_session": "MOCK"},
    )
    assert response.status_code == 200
    response_data = response.json()
    created_query_id_for_favorite_test = response_data.get("id")
    assert (
        created_query_id_for_favorite_test is not None
    ), "Query ID should not be None after creation"

    # Snapshot the created query (optional, but good for consistency)
    assert normalise_json(response_data) == snapshot


def test_toggle_query_favorite_and_list(app_server, test_db_id, snapshot):
    """Test toggling a query's favorite status and listing favorites."""
    global created_query_id_for_favorite_test
    assert (
        created_query_id_for_favorite_test is not None
    ), "Query ID for favorite test not set."

    # 1. Favorite the query
    response_fav = requests.post(
        f"{app_server}/query/{created_query_id_for_favorite_test}/favorite",
        cookies={"wos_session": "MOCK"},
    )
    assert response_fav.status_code == 200
    assert normalise_json(response_fav.json()) == snapshot(name="favorite_response")
    assert response_fav.json().get("is_favorite") is True

    # 2. List favorites and check if our query is there
    # The /favorites endpoint needs a contextId, which is typically the databaseId
    response_list_fav = requests.get(
        f"{app_server}/favorites?contextId=database-{test_db_id}",
        cookies={"wos_session": "MOCK"},
    )
    assert response_list_fav.status_code == 200
    favorites_data = normalise_json(response_list_fav.json())
    assert favorites_data == snapshot(name="list_favorites_after_favoriting")

    # Check that our specific query is in the favorites list
    assert any(
        q.get("id") == created_query_id_for_favorite_test
        for q in response_list_fav.json().get("queries", [])
    )

    # 3. Unfavorite the query
    response_unfav = requests.post(
        # Calling again toggles it off
        f"{app_server}/query/{created_query_id_for_favorite_test}/favorite",
        cookies={"wos_session": "MOCK"},
    )
    assert response_unfav.status_code == 200
    assert normalise_json(response_unfav.json()) == snapshot(name="unfavorite_response")
    assert response_unfav.json().get("is_favorite") is False

    # 4. List favorites again and check if our query is gone
    response_list_unfav = requests.get(
        f"{app_server}/favorites?contextId=database-{test_db_id}",
        cookies={"wos_session": "MOCK"},
    )
    assert response_list_unfav.status_code == 200
    favorites_data_after_unfav = normalise_json(response_list_unfav.json())
    assert favorites_data_after_unfav == snapshot(
        name="list_favorites_after_unfavoriting"
    )

    # Check that our specific query is NOT in the favorites list
    assert not any(
        q.get("id") == created_query_id_for_favorite_test
        for q in response_list_unfav.json().get("queries", [])
    )
