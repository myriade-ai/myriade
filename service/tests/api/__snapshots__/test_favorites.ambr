# serializer version: 1
# name: test_create_query_for_favorite_testing
  dict({
  })
# ---
# name: test_toggle_query_favorite_and_list[favorite_response]
  dict({
    'is_favorite': True,
    'success': True,
  })
# ---
# name: test_toggle_query_favorite_and_list[list_favorites_after_favoriting]
  dict({
    'charts': list([
    ]),
    'queries': list([
      dict({
        'completedAt': None,
        'count': None,
        'databaseId': 'a0000000-0000-0000-0000-000000000001',
        'exception': None,
        'operationType': None,
        'rows': None,
        'sql': 'SELECT 100 AS fav_col;',
        'startedAt': None,
        'status': 'completed',
        'title': 'Test Query for Favorites',
      }),
    ]),
  })
# ---
# name: test_toggle_query_favorite_and_list[list_favorites_after_unfavoriting]
  dict({
    'charts': list([
    ]),
    'queries': list([
    ]),
  })
# ---
# name: test_toggle_query_favorite_and_list[unfavorite_response]
  dict({
    'is_favorite': False,
    'success': True,
  })
# ---
